docs/node-mixin/rules/rules.libsonnet

```diff
@@ -0,0 +1,115 @@
+                count without (mode) (
+                  node_cpu_seconds_total{%(nodeExporterSelector)s}
+                )
+              )
+            ||| % $._config,
+          },
+          {
+            // CPU utilisation is % CPU is not idle.
+            record: 'instance:node_cpu_utilisation:rate1m',
+            expr: |||
+              1 - avg without (cpu, mode) (
+                rate(node_cpu_seconds_total{%(nodeExporterSelector)s, mode="idle"}[1m])
+              )
+            ||| % $._config,
+          },
+          {
+            // This is CPU saturation: 1min avg run queue length / number of CPUs.
+            // Can go over 1.
+            // TODO: There are situation where a run queue >1/core is just normal and fine.
+            //       We need to clarify how to lead this metric and if its usage is helpful at all.
+            record: 'instance:node_load1_per_cpu:ratio',
+            expr: |||
+              (
+                node_load1{%(nodeExporterSelector)s}
+              /
+                instance:node_num_cpu:sum{%(nodeExporterSelector)s}
+              )
+            ||| % $._config,
+          },
+          {
+            // Memory utilisation per node, normalized by per-node memory
+            record: 'instance:node_memory_utilisation:ratio',
+            expr: |||
+              1 - (
+                node_memory_MemAvailable_bytes{%(nodeExporterSelector)s}
+              /
+                node_memory_MemTotal_bytes{%(nodeExporterSelector)s}
+              )
+            ||| % $._config,
+          },
+          {
+            record: 'instance:node_memory_swap_io_pages:rate1m',
+            expr: |||
+              (
+                rate(node_vmstat_pgpgin{%(nodeExporterSelector)s}[1m])
+              +
+                rate(node_vmstat_pgpgout{%(nodeExporterSelector)s}[1m])
+              )
+            ||| % $._config,
+          },
+          {
+            // Disk utilisation (seconds spent, 1 second rate)
+            record: 'instance_device:node_disk_io_time_seconds:rate1m',
+            expr: |||
+              rate(node_disk_io_time_seconds_total{%(nodeExporterSelector)s, %(diskDeviceSelector)s}[1m])
+            ||| % $._config,
+          },
+          {
+            // Disk saturation (weighted seconds spent, 1 second rate)
+            record: 'instance_device:node_disk_io_time_weighted_seconds:rate1m',
+            expr: |||
+              rate(node_disk_io_time_weighted_seconds_total{%(nodeExporterSelector)s, %(diskDeviceSelector)s}[1m])
+            ||| % $._config,
+          },
+          // TODO: For the following rules, consider configurable filtering to exclude more network
+          // device names than just "lo".
+          {
+            record: 'instance:node_network_receive_bytes:rate1m',
+            expr: |||
+              sum without (device) (
+                rate(node_network_receive_bytes_total{%(nodeExporterSelector)s, device!="lo"}[1m])
```

Comment:
Why are you excluding lo? Filtering series like this is a display issue, so should be left to graphs/alerts.

Reply:
This is a recording rule meant for graphing the network traffic received by an instance. Therefore, this already aggregates over devices, and since we are not interested in the instance talking to itself, we exclude the loopback device. I don't see how not aggregating away the device would make sense for the purpose of this rule, and I don't see how including `lo` would make sense.

Reply:
The metric name `'instance:node_network_receive_bytes:rate1m` means the sum of a 1m rate of   node_network_receive_bytes_total broken down by instance. What you've here is different

Reply:
What metric name would you suggest then?

Reply:
I'd generally try to avoid this sort of thing, but something like adding `_without_lo`

Reply:
I see… That would be `instance:node_network_receive_bytes:rate1m_without_lo`? Or would you add it somewhere in the middle part of the name?

Reply:
The middle of the name

Reply:
OK. How about `instance:node_network_receive_bytes_excluding_lo:rate1m` (`excluding` rather than `without` to avoid conflation with the `without` clause in PromQL).

Reply:
sgtm

Reply:
Name change added.